### PR TITLE
Update gui test assistant

### PR DESCRIPTION
**Pull request by sjagoe**
This backports some fixes for the GuiTestAssistant to pyface.

- Avoid uncollectable garbage caused by events not being disconnected
- Avoid uncollectable garbage due to references held in `sys.last_traceback`
- Improve robustness by flushing Qt events and quitting the application on `tearDown`
- Correctly handle the case where no traits are waited for in an event loop
- Correctly handle the case where multiple traits are waited for in an event loop

### CODE REVIEW

**Comment by sjagoe** [comment]
Lots of Windows Python 2.7 failures of the form

```
======================================================================
ERROR: Failure: ImportError (No module named ui_traits)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python27-x64\lib\site-packages\nose\loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "c:\python27-x64\lib\site-packages\nose\importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "c:\python27-x64\lib\site-packages\nose\importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "c:\python27-x64\lib\site-packages\pyface\action\tests\test_action.py", line 5, in <module>
    from ..action import Action
  File "c:\python27-x64\lib\site-packages\pyface\action\action.py", line 20, in <module>
    from traitsui.ui_traits import Image
  File "c:\python27-x64\lib\site-packages\traitsui\ui_traits.py", line 27, in <module>
    from pyface.ui_traits import (Alignment, Border, HasBorder, HasMargin, Image,
ImportError: No module named ui_traits
```

**Comment by corranwebster** [comment]
Something is weird here - it looks like Appveyor is picking up an old version of Pyface (eg. last release) rather than master, but running traitsui master (which is correct).  `pyface.ui_traits` certainly exists in master: https://github.com/enthought/pyface/blob/master/pyface/ui_traits.py

**Comment by corranwebster** [comment]
Even more so: the line 20 in the `pyface.actions.action` is https://github.com/enthought/pyface/blob/master/pyface/action/action.py#L21 in current master and is importing directly from Pyfce, as it should.  So I think something is wrong with the Appveyor setup for Python 2.7.

**Comment by dpinte** [comment]
@corranwebster the appveyor configuration uses `pip` configured to use `c:\temp` as cache directory, which is cached by appveyor. Could it be an issue with the way appveyor does the caching?

**Comment by corranwebster** [comment]
Maybe, I'm not familiar enough with the Appveyor setup - the current entangled dependencies of TraitsUI and Pyface mean that doing `setup.py` for TraitsUI will likely install a master of Pyface; which would normally be OK, because we immediately replace that with the master of Pyface.  To see what we're seeing the install of the master of Pyface would have to fail or otherwise not be picked up (eg. bad `PYTHONPATH`, etc.)

The correct fix is probably to have Appveyor run the EDM-based tests that we have working on Travis CI.

**Comment by mdickinson** [comment]
I'm testing these changes against downstream projects; will report back.

**Comment by mdickinson** [comment]
Looking good so far. I have a couple of extra tests in the downstream project that I can push to this branch, if that's okay with everyone (if not, I can create a PR against this branch).

**Comment by itziakos** [comment]
> Could it be an issue with the way appveyor does the caching?

Traits and TraitsUI are installed directly from the github repo, pip will always install the cloned repo so there is no caching taking place. Could it be a circular import issue?

File: pyface/ui/qt4/util/gui_test_assistant.py [hunk 1]
```diff
@@ -12,8 +12,8 @@
 
 import mock
 
+from pyface.gui import GUI
```
Comment (itziakos) [inline pyface/ui/qt4/util/gui_test_assistant.py]: Please revert this import to the original importing from `qt4`. Importing from `pyface.gui` from inside the `ui.qt4` submodule is not right (and probably not totally safe)

**Comment by sjagoe** [comment]
@mdickinson 

> that I can push to this branch

You can go ahead and push your extra tests directly here.

Thanks.

File: pyface/ui/qt4/util/gui_test_assistant.py (hunk 1 above)
Comment (sjagoe) [inline pyface/ui/qt4/util/gui_test_assistant.py]: True; I didn't notice that change when pulling the updates in.

**Comment by corranwebster** [comment]
@itziakos with a little more digging: so it looks like with the appveyor setup, you get two versions of Pyface:

* a Pyface 5.1 installed by the pip install traitsui command, pulled from PyPI and installed into `site-packages`.
* the Pyface installed using `python setup.py develop` (so version is 5.2.0dev0).

The latter should be what you get when you import from pyface, but it appears that for some reason the `nose` importer is picking up the Pyface 5.1, not the 5.2.0dev0.  So either the `python setup.py develop` is failing in some way, or the nose importer (or something more fundamental) is ignoring the information about python paths installed by the setup.py develop.

No idea why this is working on 3.5 but not 2.7.

**Comment by corranwebster** [comment]
And for additional fun, we seem to be getting the occasional OS X + PyQt segfaults here as well.

**Comment by corranwebster** [comment]
Marged master should fix issues with Appveyor tests.

**Comment by mdickinson** [comment]
Ready to merge?

**Comment by corranwebster** [comment]
I'm happy with it.

**Pull request merged**